src/java.base/share/classes/jdk/internal/foreign/MemoryInspectionUtil.java

```diff
@@ -0,0 +1,293 @@
+            return;
+        }
+        if (layout instanceof SequenceLayout sequenceLayout) {
+            action.accept(indentedLabel(state, sequenceLayout) + " [");
+            state.incrementIndent();
+            final long elementCount = sequenceLayout.elementCount();
+            for (long i = 0; i < elementCount; i++) {
+                toString0(segment, sequenceLayout.elementLayout(), renderer, action, state, (i != (elementCount - 1L)) ? "," : "");
+            }
+            state.decrementIndent();
+            action.accept(state.indentSpaces() + "]" + suffix);
+            return;
+        }
+        action.accept(state.indentSpaces() + "Unknown layout: " + layout + " at index " + state.index());
+        state.indexAndAdd(layout);
+    }
+
+    static String renderValueLayout(ViewState state,
+                                    ValueLayout layout,
+                                    String value,
+                                    String suffix) {
+        return indentedLabel(state, layout) + "=" + value + suffix;
+    }
+
+    static String indentedLabel(ViewState state,
+                                MemoryLayout layout) {
+        return state.indentSpaces() + layout.name()
+                .orElseGet(layout::toString);
+    }
+
+    static final class ViewState {
+
+        private static final int SPACES_PER_INDENT = 4;
+
+        // Holding a non-static indents allows simple thread-safe use
+        private final StringBuilder indents = new StringBuilder();
+
+        private int indent;
+        private long index;
+
+        void incrementIndent() {
+            indent++;
+        }
+
+        void decrementIndent() {
+            indent--;
+        }
+
+        String indentSpaces() {
+            final int spaces = indent * SPACES_PER_INDENT;
+            while (indents.length() < spaces) {
+                // Expand as needed
+                indents.append(" ");
+            }
+            return indents.substring(0, spaces);
+        }
+
+        long index() {
+            return index;
+        }
+
+        void index(long index) {
+            this.index = index;
+        }
+
+        long indexAndAdd(long delta) {
+            final long val = index;
+            index += delta;
+            return val;
+        }
+
+        long indexAndAdd(MemoryLayout layout) {
+            return indexAndAdd(layout.byteSize());
+        }
+    }
+
+    private static final class StandardValueLayoutRenderer implements MemoryInspection.ValueLayoutRenderer {
+        @Override
+        public String toString() {
+            return singletonToString(StandardValueLayoutRenderer.class);
+        }
+    }
+
+    private static String singletonToString(Class<?> implementingClass) {
+        return "The " + implementingClass.getName() + " singleton";
+    }
+
+    static final class SingleFunctionValueLayoutRenderer implements MemoryInspection.ValueLayoutRenderer {
```

Comment:
As we discussed offline, perhaps using a BiFunction might be good enough for all practical purposes